### PR TITLE
Add support for customising the creation of Kestrel listen sockets

### LINKED ISSUE #32794: Configure the Accept Socket
This issue is inspired by @deanward81's post at https://bakedbean.org.uk/posts/2021-05-airdrop-anywhere-part-3/

### Is your feature request related to a problem? Please describe.

We recently exposed the Listen socket via a connection feature. In similar vein, we should probably also expose the Listen Socket via a server feature.

The workaround today is to replace the TransportFactory (code snippet copied from the blog post)

```csharp
// Actual implementation: https://github.com/deanward81/AirDropAnywhere/blob/2021-05-17-receiving-files/src/AirDropAnywhere.Core/HttpTransport/AwdlSocketTransportFactory.cs
internal class AwdlSocketTransportFactory : IConnectionListenerFactory
{
    private readonly IConnectionListenerFactory _connectionListenerFactory;
    private readonly FieldInfo _listenSocketField;
    
    public AwdlSocketTransportFactory(IOptions<SocketTransportOptions> options, ILoggerFactory loggerFactory)
    {
        _connectionListenerFactory = new SocketTransportFactory(options, loggerFactory);
        // HACK: this merry little reflective dance is because of sealed internal classes
        // and no extensibility points, yay :/
        _listenSocketField = typeof(SocketTransportFactory).Assembly
            .GetType("Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener")
            .GetField("_listenSocket", BindingFlags.Instance | BindingFlags.NonPublic);
    }

    public async ValueTask<IConnectionListener> BindAsync(EndPoint endpoint, CancellationToken cancellationToken = default)
    {
        var transport = await _connectionListenerFactory.BindAsync(endpoint, cancellationToken);
        // HACK: fix up the listen socket to support listening on AWDL
        var listenSocket = (Socket?) _listenSocketField.GetValue(transport);
        if (listenSocket != null)
        {
            listenSocket.SetAwdlSocketOption();
        }
        return transport;
    }
}
```

### Describe the solution you'd like

I haven't thought this through yet. 😅 I know an `IStartupFilter` is too early since the server hasn't started yet 

### Additional context

Here's a fun read: https://bakedbean.org.uk/posts/2021-05-airdrop-anywhere-part-3/

### PR DESCRIPTION
**Pull request by deanward81**
Fixes #32794. As mentioned in the issue it is currently not possible to configure the underlying listen socket used by Kestrel. In certain circumstances it is desirable to be able to configure socket options when the socket is created - a concrete case that I recently came across is setting the `SO_RECV_ANYIF` socket option on macOS so that Kestrel can listen on the `awdl0` interface (more details on that specific case [here](https://bakedbean.org.uk/posts/2021-05-airdrop-anywhere-part-3/)).

This change adds the API suggested by #32794 by adding a function that allows the caller to manage the creation of the listening socket - `Func<EndPoint, Socket> CreateListenSocket { get; set; }` allows the creation of the listen socket to be configured.

There's also some tests using IPv4, IPv6, unix domain sockets and file handles.

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:78]: Would `ConfigureAcceptedSocket` be clearer?

Comment (deanward81) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:78]: @Tratcher no objections here, will update it tomorrow morning 👍

**Comment by halter73** [comment]
I wonder if it wouldn't be better to change these from `Action<EndPoint, Socket>` to `Func<EndPoint, Socket>`.  If we went this route, we'd want to expose a `Socket CreateSocket(EndPpoint)` method somewhere and document it with these callbacks.

**Comment by Tratcher** [comment]
Create seems like a different scenario from augmentation, and requires more from the implementer.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:117]: If we go the `Action<EndPoint, Socket>` route, wouldn't it be better to call it after binding but before listening? This would make the FileHandleEndPoint and IPEndPoint cases  more similar.

If we go the `Func<EndPoint, Socket>` route, I'd expect the returned Socket to already be bound.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:76]:
```suggestion
        /// An action used to configure an accepted socket before it's used.
```

**Comment by halter73** [comment]
> and requires more from the implementer.

Not really given `Socket CreateListenSocket(EndPoint)` exists.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (JamesNK) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:71]:
```suggestion
        /// An action used to configure the listening socket before it's bound.
```

**Comment by halter73** [comment]
New proposed API discussed in triage:

```diff
public class SocketTransportOptions
{
+    public Func<EndPoint, Socket> CreateListenSocket { get; set; }
+    public Socket CreateDefaultListenSocket(EndPoint);
}
```

We don't think `ConfigureAcceptedSocket` adds much value. We recommend adding [connection middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0#connection-middleware) and using the new `IConnectionSocketFeature` (https://github.com/dotnet/aspnetcore/pull/31588).

**Comment by davidfowl** [comment]
@halter73 how do you "call base" for the listen socket.

**Comment by Tratcher** [comment]
Is CreateDefaultListenSocket supposed to be static?

**Comment by davidfowl** [comment]
Yea it should be the stateless default.

**Comment by halter73** [comment]
Yeah. It should be static. I think it would basically be the current implementation [SocketConnectionListen.Bind()](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs#L87-L140), but instead of calling `Socket.Listen(int)` and setting `_listenSocket`, it would return the Socket before listening. The new implementation of `Bind()` would then call the `SocketTransportOptions.CreateListenSocket(EndPoint)` Func and then call `Socket.Listen(int)` on the returned value and set it to `_listenSocket`.

**Comment by halter73** [comment]
If we make it non-static, we *could* do some trickery where the SocketConnectionListener adds an internal reference to itself to SocketTransportOptions before calling the CreateDefaultListenSocket callback allowing CreateDefaultListenSocket to set `_socketHandle` if need be. Not sure if this would be worthwhile, but I do like that it makes it work in all cases.

**Comment by deanward81** [comment]
> We might have to bypass the CreateListenSocket for FileHandleEndPoints since I can't think of a great way to manage the `_socketHandle` in that case. I think this is fine because if you're using a FileHandleEndPoint, you can configure the socket options before passing it to `KestrelServerOptions.ListenHandle(long)`.

Does the `_socketHandle` field need to be there? It's only used during disposal, but that handle is created with `ownsHandle: true` which, when passed to `Socket`, means the handle is cleaned up in the finaliser of the socket (which arguably is more correct because it's an unmanaged resource). https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3262

That said, is the handle _really_ owned in this case? The user is passing an opaque handle created elsewhere - is it intentional that it is marked as "owned" by the socket transport? Shouldn't they be responsible for cleanup?

Reason I ask - not having that field allows `CreateDefaultListenSocket` to work for all cases and keeps things decoupled (no need to do mess with references to the `SocketConnectionListener` in the instance case).

**Comment by deanward81** [comment]
I've updated to use the triaged API (using a static `CreateDefaultListenSocket`). I removed the `_listenHandle` in the last commit and, so far as I can tell, there are no negative consequences (tests pass, at least). Details (from the commit):

> Note that during implementation I removed the `_socketHandle` field from `SocketConnectionListener` - this was only set so that `Dispose` could be called when the listener is disposed. Under the hood `Socket` already disposes a handle passed to it during finalization, but only if the `ownsHandle` parameter is `true` . In this case the `SafeSocketHandle` _is_ instantiated with  this parameter so the the underlying handle will be closed when the `_listenSocket` field is disposed - that is currently the case when the listener is disposed.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:26]: I think it's fine to remove this since we probably shouldn't have taken ownership of the handle in the first place, but I want to point out to other reviewers this is a behavioral change.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:75]: Can we preinitialize this with CreateDefaultListenSocket and make it non-nullable?

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:93]:
```suggestion
                        new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: false)
```

Since we're not managing the lifetime anymore.

Comment (deanward81) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:93]: Interestingly, when I do this the test case that uses a `FileHandleEndPoint` never returns - the shutdown of Kestrel hangs indefinitely (when calling `_transportManager.StopAsync()`) - presumably because the underlying socket isn't closed. In the case of the test this is because the underlying `Socket` I steal the handle from is static and I'm unable to manage its lifetime as it's consumed using `MemberData`.

I can absolutely fix the test case but I'm now dubious to change it because it's possible that Kestrel might hang if the caller doesn't clean-up the handle that they supposedly own. That seems like a bad place to be without much indication what the problem is.

Comment (deanward81) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:93]: Thinking that adding a comment and leaving the `ownsHandle: true` might be a safer route to go here

Comment (halter73) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:93]: This makes me think we should just go the `Func<EndPoint, Func<EndPoint, Socket>, Socket>` route since it's the most-flexible/least-tricky way to do this while continuing to manage `_socketHandle` lifetime. This is a pretty advanced API anyway. I don't like the idea of SafeHandle's finalizer controlling the lifetime of the socket.

**Comment by halter73** [comment]
New new proposed API based on https://github.com/dotnet/aspnetcore/pull/32827#discussion_r636519952

public class SocketTransportOptions
{
+    public Func<EndPoint, Func<EndPoint, Socket>, Socket> CreateListenSocket { get; set; }
}

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs
Comment (halter73) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:93]: Actually, it looks like Socket itself disposes the SafeSocketHandle if we claim ownership. It seems like we were doing a bunch of work for nothing. I think your change is good as is. Maybe just add a comment saying that `ownsHandle` indicates the [Socket itself will dispose the handle when it is Disposed](https://github.com/dotnet/runtime/blob/999c9c0bf1162e624533e2df58062d2811961fd3/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3270).

**Review by halter73 — APPROVED** [review]
I now see that the SafeSocketHandle is disposed by Socket all along because we were specifying `ownsHandle: true`. This means the previous manual tracking of the SafeSocketHandle was redundant and there's no behavioral change. Or at least that's what it seems like. I'll step through the code to confirm before merging.

**Comment by JamesNK** [comment]
@halter73 Static method is new as well for API review?

**Comment by halter73** [comment]
Yes. This is still `api-ready-for-review`. We shouldn't merge before it's approved.

**Comment by BrennanConroy** [comment]
API review:
```
public static Socket CreateDefaultBoundListenSocket(EndPoint);
public Func<EndPoint, Socket> CreateBoundListenSocket { get; set; }
```

**Comment by deanward81** [comment]
@BrennanConroy currently the implementation of `CreateDefaultListenSocket` has not had `Bind` called on it - with the approved API it implies that is the case - presumably the implementation should be changed to do so?

**Comment by deanward81** [comment]
I've updated the code to match the API review. Let me know if you want any additional tweaks @BrennanConroy

**Comment by davidfowl** [comment]
@BrennanConroy @halter73 is this good?

**Comment by halter73** [comment]
Thanks @deanward81!

**Comment by sebastienros** [comment]
Did we measure the perf impact of that change? Asking for a friend :/

**Comment by halter73** [comment]
@sebastienros We didn't measure the perf of this change, but I'd be shocked if this had a perf impact. This moves the creation of the *listen* socket (not the accepted socket) to a Func, but keeps the logic the same by default. Even if there was a perf hit from calling a Func, it's only called once per address at app startup.

**Comment by halter73** [comment]
#34769 would be more likely to affect `Connection: close` perf because it touches a per-connection code path. I don't think we explicitly perf tested that. However, it's basically just a refactoring, so it shouldn't impact perf.

**Pull request merged**
